manpage/arjun.1
.\" Text automatically generated by txt2man
.TH arjun 1 "14 Jan 2023" "arjun-2.2.1" "HTTP parameter discovery suite"
.SH NAME
\fBarjun \fP- HTTP parameter discovery suite
\fB
.SH SYNOPSIS
.nf
.fam C
\fBarjun\fP [\fB-h\fP] [\fB-u\fP \fIURL\fP] [\fB-o\fP \fIJSON_FILE\fP] [\fB-oT\fP \fITEXT_FILE\fP] [\fB-oB\fP [\fIBURP_PORT\fP]] [\fB-d\fP \fIDELAY\fP] [\fB-t\fP \fITHREADS\fP] [\fB-w\fP \fIWORDLIST\fP] [\fB-m\fP \fIMETHOD\fP] [\fB-i\fP [\fIIMPORT_FILE\fP]]
      [\fB-T\fP \fITIMEOUT\fP] [\fB-c\fP \fICHUNKS\fP] [\fB-q\fP] [\fB--headers\fP [\fIHEADERS\fP]] [\fB--passive\fP [\fIPASSIVE\fP]] [\fB--stable\fP] [\fB--include\fP \fIINCLUDE\fP] [\fB--disable-redirects\fP]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
Arjun can find query parameters for \fIURL\fP endpoints. If you don't get what that means, it's okay, read along.
Web applications use parameters (or queries) to accept user input, take the following example into consideration:
.PP
.nf
.fam C
    http://api.example.com/v1/userinfo?id=751634589

.fam T
.fi
This \fIURL\fP seems to load user information for a specific user id, but what if there exists a parameter named admin which
when set to True makes the endpoint provide more information about the user?
This is what Arjun does, it finds valid HTTP parameters with a huge default dictionary of 25,890 parameter names.
The best part? It takes less than 10 seconds to go through this huge list while making just 50-60 requests to the target. Here's how.
.TP
.B
-
Supports GET/POST/POST-JSON/POST-XML requests
.TP
.B
-
Automatically handles rate limits and timeouts
.TP
.B
-
Export results to: BurpSuite, text or JSON file
.TP
.B
-
Import targets from: BurpSuite, text file or a raw request file
.TP
.B
-
Can passively extract parameters from JS or 3 external sources
.SH OPTIONS
.TP
.B
\fB-h\fP, \fB--help\fP
show this help message and exit.
.TP
.B
\fB-u\fP \fIURL\fP
Target \fIURL\fP.
.TP
.B
\fB-o\fP \fIJSON_FILE\fP, \fB-oJ\fP \fIJSON_FILE\fP
Path for json output file.
.TP
.B
\fB-oT\fP \fITEXT_FILE\fP
Path for text output file.
.TP
.B
\fB-oB\fP [\fIBURP_PORT\fP]
Port for output to Burp Suite Proxy. Default port is 8080.
.TP
.B
\fB-d\fP \fIDELAY\fP
Delay between requests in seconds. (default: 0).
.TP
.B
\fB-t\fP \fITHREADS\fP
Number of concurrent threads. (default: 5).
.TP
.B
\fB-w\fP \fIWORDLIST\fP
Wordlist file path. (default: /usr/lib/python3/dist-packages/\fBarjun\fP/db/large.txt).
.TP
.B
\fB-m\fP \fIMETHOD\fP
Request method to use: GET/POST/XML/JSON/\fIHEADERS\fP. (default: GET).
.TP
.B
\fB-i\fP [\fIIMPORT_FILE\fP]
Import target URLs from file.
.TP
.B
\fB-T\fP \fITIMEOUT\fP
HTTP request timeout in seconds. (default: 15).
.TP
.B
\fB-c\fP \fICHUNKS\fP
Chunk size. The number of parameters to be sent at once.
.TP
.B
\fB-q\fP
Quiet mode. No output.
.TP
.B
\fB--headers\fP [\fIHEADERS\fP]
Add headers. Separate multiple headers with a new line.
.TP
.B
\fB--passive\fP [\fIPASSIVE\fP]
Collect parameter names from passive sources like wayback, commoncrawl and otx.
.TP
.B
\fB--stable\fP
Prefer stability over speed.
.TP
.B
\fB--include\fP \fIINCLUDE\fP
Include this data in every request.
.TP
.B
\fB--disable-redirects\fP
disable redirects.
.SH EXAMPLES
\fBarjun\fP \fB-h\fP
.PP
\fBarjun\fP \fB-u\fP http://site.example.com/test.php
.PP
\fBarjun\fP \fB-u\fP http://site.example.com/test.php \fB-o\fP test.json
.PP
\fBarjun\fP \fB-u\fP http://site.example.com \fB-t\fP 5
.PP
\fBarjun\fP \fB-u\fP http://site.example.com \fB--stable\fP
.RE
.PP

.SH AUTHOR
Written by Somdev Sangwan <s0md3v@gmail.com>
.PP
This manual page was written by Guilherme de Paula Xavier Segundo
<guilherme.lnx@gmail.com> for the Debian project (but may be used by others).


CHANGELOG.md
#### 2.2.0
- Ability to detect parameters that respond to a certain value e.g. "?debug=yes"
- Added "required parameter" detection 
- Heuristic can now extract words out of json/text responses
- Fixed -oB option description

#### 2.1.6
- Fixed multiple breaking bugs
- Export results as they come in multi-target mode
- Various improvements to output in multi-target mode
- changed default chunk size 300->500 and threads to 2->5

#### 2.1.5
- Fixed header comparison (will fix infinite bruteforce on some targets)
- Fixed catastrophic backtracking in some regexes (arjun used to get stuck)
- New logic for handling redirections
- `--disable-redirects` option

#### 2.1.4
- Fixed file-paths not being windows compatible
- Fixed and improved JavaScript heuristics scanning
- Fixed missing ampersands in `-oT` output
- Refactoring of help options and code

#### 2.1.3
- Fixed memory exhaustion bug
- Fixed parsing of raw HTTP files
- Added new detection factor: `number of lines`
- Failed retries are now handled properly

#### 2.1.2
- Minor code cleanup
- Fixed `--headers` option

#### 2.1.1
- Fixed circular import
- Fixed BurpSuite export
- Fixed not working headers
- Better response type checking
- Fixed wordlist error on Windows
- Fixed `Content-Type` header bug

#### 2.1.0
- Added `XML` method
- `-q` option for quiet mode
- New wordlists backed by research
- `-oT` option for txt export
- `-oB` option for BurpSuite export
- `-oJ` alias for JSON export
- Added support for custom injection point in `XML` and `JSON`
- pypi package

#### 2.0-beta
- Added an anomaly detection algorithm with 9 factors
- Added a HTTP response analyzer for handling errors and retrying requests
- Significantly improved heuristic scanner
- `--passive` option for collecting parameters from otx, commoncrawl and archive.org
- `-c` option to define number of parameters to be sent at once
- import via `-i` options now supports: BurpSuite log, raw request file, text file with urls
- `-T` option to specify HTTP connection timeout
- combined `-m` option for specifying HTTP request method
- Various bug fixes and better output

#### 1.6
- `--stable` switch for handling rate limits
- Include empty JS variables for testing
- Various optimizations and bug fixes
- Handle keyboard interruption
- Removed redundant code

#### 1.5
- Ignore dynamic content
- Detect int-only parameters
- Include URL in json output
- Track each reflection separately
- Improved error handling

#### 1.4
- Added `JSON` support
- Fixed a major bug in detection logic
- `-o` option to save result to a file
- `--urls` option to scan list of URLs
- Ability to supply HTTP headers from CLI

#### 1.3
- improved logic
- detection by plain-text content matching
- `--include` switch to include persistent data
- fixed a bug that caused user supplied HTTP headers to have no effect

#### 1.2-beta
- Drastic performance improvement (x50 faster)

#### 1.1
Initial stable release


arjun.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import print_function

from core.colors import green, end, info, bad, good, run, res

print('''%s    _
   /_| _ '
  (  |/ /(//) v2.0-beta
      _/      %s
''' % (green, end))

try:
    from concurrent.futures import ThreadPoolExecutor, as_completed
except ImportError:
    print('%s Please use Python > 3.2 to run Arjun.' % bad)
    quit()

import sys
import json
import time
import argparse

from itertools import repeat
from urllib.parse import urlparse

import core.config as mem
from core.bruter import bruter
from core.prompt import prompt
from core.importer import importer
from core.requester import requester
from core.anamoly import compare, define
from core.utils import fetch_params, stable_request, randomString, slicer, confirm, getParams, populate, removeTags, extractHeaders, parse_request, reader
from jsonmerge import merge

from plugins.heuristic import heuristic

parser = argparse.ArgumentParser() # defines the parser
# Arguments that can be supplied
parser.add_argument('-u', help='target url', dest='url')
parser.add_argument('-o', help='path for the output file', dest='output_file')
parser.add_argument('-d', help='delay between requests', dest='delay', type=float, default=0)
parser.add_argument('-t', help='number of threads', dest='threads', type=int, default=2)
parser.add_argument('-w', help='wordlist path', dest='wordlist', default=sys.path[0]+'/db/params.txt')
parser.add_argument('-m', help='request method: GET/POST/JSON', dest='method', default='GET')
parser.add_argument('-i', help='import targets from file', dest='import_file', nargs='?', const=True)
parser.add_argument('-T', help='http request timeout', dest='timeout', type=float, default=15)
parser.add_argument('-c', help='chunk size/number of parameters to be sent at once', type=int, dest='chunks', default=500)
parser.add_argument('--headers', help='add headers', dest='headers', nargs='?', const=True)
parser.add_argument('--passive', help='collect parameter names from passive sources', dest='passive')
parser.add_argument('--stable', help='prefer stability over speed', dest='stable', action='store_true')
parser.add_argument('--include', help='include this data in every request', dest='include', default={})
args = parser.parse_args() # arguments to be parsed

mem.var = vars(args)

mem.var['method'] = mem.var['method'].upper()

if mem.var['stable'] or mem.var['delay']:
    mem.var['threads'] = 1

try:
    wordlist = set(reader(args.wordlist, mode='lines'))
    if mem.var['passive']:
        host = mem.var['passive']
        if host == '-':
            host = urlparse(args.url).netloc
        print('%s Collecting parameter names from passive sources for %s, it may take a while' % (run, host))
        passive_params = fetch_params(host)
        wordlist.update(passive_params)
        print('%s Collected %s parameters, added to the wordlist' % (info, len(passive_params)))
    wordlist = list(wordlist)
except FileNotFoundError:
    exit('%s The specified file for parameters doesn\'t exist' % bad)

if not (args.url, args.import_file):
    exit('%s No targets specified' % bad)

def prepare_requests(args):
    headers = {
    'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:83.0) Gecko/20100101 Firefox/83.0',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.5',
    'Accept-Encoding': 'gzip, deflate',
    'Connection': 'close',
    'Upgrade-Insecure-Requests': '1'
    }
    if type(headers) == bool:
        headers = extractHeaders(prompt())
    elif type(headers) == str:
        headers = extractHeaders(headers)
    if mem.var['method'] == 'JSON':
        mem.headers['Content-type'] = 'application/json'
    if args.url:
        params = getParams(args.include)
        return {
            'url': args.url,
            'method': mem.var['method'],
            'headers': headers,
            'include': params
        }
    elif args.import_file:
        return importer(args.import_file, mem.var['method'], headers, args.include)
    return []


def narrower(request, factors, param_groups):
    anamolous_params = []
    threadpool = ThreadPoolExecutor(max_workers=mem.var['threads'])
    futures = (threadpool.submit(bruter, request, factors, params) for params in param_groups)
    for i, result in enumerate(as_completed(futures)):
        if result.result():
            anamolous_params.extend(slicer(result.result()))
        if not mem.var['kill']:
            print('%s Processing chunks: %i/%-6i' % (info, i + 1, len(param_groups)), end='\r')
    return anamolous_params

def initialize(request, wordlist):
    url = request['url']
    if not url.startswith('http'):
        print('%s %s is not a valid URL' % (bad, url))
        return 'skipped'
    print('%s Probing the target for stability' % run)
    stable = stable_request(url, request['headers'])
    if not stable:
        return 'skipped'
    else:
        fuzz = randomString(6)
        response_1 = requester(request, {fuzz : fuzz[::-1]})
        print('%s Analysing HTTP response for anamolies' % run)
        fuzz = randomString(6)
        response_2 = requester(request, {fuzz : fuzz[::-1]})
        if type(response_1) == str or type(response_2) == str:
            return 'skipped'
        factors = define(response_1, response_2, fuzz, fuzz[::-1], wordlist)
        print('%s Analysing HTTP response for potential parameter names' % run)
        found = heuristic(response_1.text, wordlist)
        if found:
            num = len(found)
            s = 's' if num > 1 else ''
            print('%s Heuristic scanner found %i parameter%s: %s' % (good, num, s, ', '.join(found)))
        print('%s Logicforcing the URL endpoint' % run)
        populated = populate(wordlist)
        param_groups = slicer(populated, int(len(wordlist)/args.chunks))
        last_params = []
        while True:
            param_groups = narrower(request, factors, param_groups)
            if mem.var['kill']:
                return 'skipped'
            param_groups = confirm(param_groups, last_params)
            if not param_groups:
                break
        confirmed_params = []
        for param in last_params:
            reason = bruter(request, factors, param, mode='verify')
            if reason:
                name = list(param.keys())[0]
                confirmed_params.append(name)
                print('%s name: %s, factor: %s' % (res, name, reason))
        return confirmed_params

request = prepare_requests(args)

final_result = {}

try:
    if type(request) == dict:
        mem.var['kill'] = False
        url = request['url']
        these_params = initialize(request, wordlist)
        if these_params == 'skipped':
            print('%s Skipped %s due to errors' % (bad, request['url']))
        elif these_params:
            final_result['url'] = url
            final_result['params'] = these_params
            final_result['method'] = request['method']
    elif type(request) == list:
        for each in request:
            final_result={}
            url = each['url']
            mem.var['kill'] = False
            print('%s Scanning: %s' % (run, url))
            these_params = initialize(each, list(wordlist))
            if these_params == 'skipped':
                print('%s Skipped %s due to errors' % (bad, url))
            elif these_params:
                final_result[url] = {}
                final_result[url]['params'] = these_params
                final_result[url]['method'] = each['method']
                print('%s Parameters found: %s' % (good, ', '.join(final_result[url])))
            if args.output_file and final_result:
                try:
                    with open(str(mem.var['output_file']), 'r', encoding='utf8') as json_input:
                        inp=json.loads(json_input.read())
                except:                         #Happens first time when the file is empty
                        inp={}                      
                final=merge(inp,final_result) 
                with open(str(mem.var['output_file']), 'w+', encoding='utf8') as json_output:
                    json.dump(final, json_output, sort_keys=True, indent=4)
                print('%s Output saved to JSON file in %s' % (info, mem.var['output_file']))

except KeyboardInterrupt:
    exit()
